### ISSUE TITLE
Some issues when plotting `bipc` layout network

### ISSUE THREAD

**Issue by tapiocaaaa** [mplex/multigraph#5]
Hi all,
I am a beginner in the R programming language. I am trying to plot a bipartite network using `multigraph` package. But, I am not familiar with it, and I encountered some issues while plotting.

The graph below was plotted using `bip3` layout. I would like to know if it is possible to determine the number of dots in each columns through setting arguments. For example, I wanna the first 21 dots are set to be on the left side and rest of the dots are on the right. Additionally, I used 'clu' to differentiate each cluster, but the result was not what I expected. I would like to separate three columns using different colors, so column one in red, middle column in green and the right one in blue will be ideal.

It will be very appreciated that if someone can help!
Thanks a lot!

<img width="501" alt="截屏2024-03-11 10 55 56" src="https://github.com/mplex/multigraph/assets/161675555/d3cfa509-f214-4feb-b188-25243a456bf0">

**Comment by tapiocaaaa** [comment]
My data is a three dimensional array:
`total_adj<-array(0,c(34,14,2),dimnames = list(total_dim1,total_dim2,total_dim3))`

I tried `bipc` layout to plot my network
`r<-c(rep(1,21),rep(2,13))`
`c<-c(rep(1,14))`
`bmgraph(total_adj,layout = "bipc",clu = list(r,c))`
The code resulted in an error: 
Error in net[cls1, cls2] : incorrect number of dimensions
When I plotted the network using one dimension data, it worked. The first 21 dots are on the left.
`bmgraph(total_adj[,,1],layout = "bipc",clu = list(r,c))`
<img width="487" alt="截屏2024-03-11 15 17 22" src="https://github.com/mplex/multigraph/assets/161675555/c6c9cc17-7a14-438b-ba11-14ab57ee7e2e">
But I need 3 dimensional data to present the blue edges and red edges. And I still have not figured out how to separate columns using different colors.
It will be very appreciated that if someone can help!
Thanks a lot!

**Comment by mplex** [comment]
> ... I would like to know if it is possible to determine the number of dots in each columns through setting arguments. For example, I wanna the first 21 dots are set to be on the left side and rest of the dots are on the right. Additionally, I used 'clu' to differentiate each cluster, but the result was not what I expected. I would like to separate three columns using different colors, so column one in red, middle column in green and the right one in blue will be ideal.
> ...
> <img alt="截屏2024-03-11 10 55 56" width="501" src="https://private-user-images.githubusercontent.com/161675555/311576706-d3cfa509-f214-4feb-b188-25243a456bf0.png?
> === TRUNCATED ===

<br/>

Thanks for the questions.
<br/>

> *issue: "I wanna the first 21 dots are set to be on the left side and rest of the dots are on the right"*

Automatic layouts with `bmgraph` will split the data in equal parts (nearly if *n* is odd number).
More flexibility you will find by using a customized coordinate system using argument `coord` instead of the `layout` argument.

A couple of internal functions from `multigraph` can help the coordinated system: 

* `rng` gives you a range of *n* elements
* `nrm` "normalizes" any range from `0` to `1`

Example for *n*=7 with a native pipe operator:

```
rng(7)
#[1] -0.14000000 -0.09333333 -0.04666667  0.00000000  0.04666667  0.09333333  0.14000000

rng(7) |> nrm()
#[1] 0.000 0.167 0.333 0.500 0.667 0.833 1.000
```



<br/>

> *issue: "separate three columns using different colors"*

Argument `vclu` is meant for color clustering of the vertices in the two domains of the affiliation network where the colors are specified in `vcol`. That is:

```
array(0, c(17*2,(48-(17*2))), dimnames=list(1:34,1:14)) |> 
  bmgraph(layout="bip3", vclu=list(c(rep(1,17),rep(2,17)),c(rep(3,14))), vcol=3:5, cex=3)
```

![billede](https://github.com/mplex/multigraph/assets/20990910/ac988dcc-e721-4139-8f0c-9d09b901d355)

<br/>

> *error issue*

> I tried bipc layout to plot my network
> r<-c(rep(1,21),rep(2,13))
> c<-c(rep(1,14))
> bmgraph(total_adj,layout = "bipc",clu = list(r,c))
> The code resulted in an error:
> Error in net[cls1, cls2] : incorrect number of dimensions

<br/>
Thanks for reporting this issue.

Layout `bipc` did not support three-dimensional arrays.
There is now a new version of function `bmgraph` on GitHub that fix this bug, and `v0.99-2` of `multigraph` package will be part of the new submission to CRAN repository.

<br/>

Hope it helps.
